Add pipeline snippets for Firestore

morganchen12 opened the pull request.

morganchen12 committed:
add initial pipelines snippet file

morganchen12 committed:
add pipeline snippets for python

daniel-sanche commented on snippets/firestore/firestore_pipelines.py:
```diff
@@ -0,0 +1,1289 @@
+        .select(
+            Field.of("current").logical_minimum(Field.of("updated")).as_("price_min")
+        ) \
+        .execute()
```
nit: I would probably merge this with the above line

daniel-sanche commented on snippets/firestore/firestore_pipelines.py:
```diff
@@ -0,0 +1,1289 @@
+    for res in result:
+        print(res)
+
+def sqrt_function():
```
This seems like a duplicate of the above? If that's really what you want, I think you can stack multiple region tags over the same sample

daniel-sanche replied:
My understanding is no, and I've definitely seen other samples that double up like that. (Here's [an example](https://github.com/googleapis/python-logging/blob/a7a28d1b9396e472f153fd76d0d846338c107d5a/samples/snippets/handler.py#L20))

But if you want to keep them separate, that's fine

daniel-sanche commented on snippets/firestore/firestore_pipelines.py:
```diff
@@ -0,0 +1,1289 @@
+    result = client.pipeline() \
+        .collection("books") \
+        .select(
+            And(
```
Maybe you should include the import in the sample? `from google.cloud.firestore_v1.pipeline_expressions import And`. I guess this feedback would apply to Field and Constant too though...

Our usual policy was that samples should all include relevant imports. Most of these pipeline tests don't really need imports. But when we start adding Ands and Ors and Nots, I could see users getting confused, since they look so close to built-in methods

daniel-sanche commented on snippets/firestore/firestore_pipelines.py:
```diff
@@ -0,0 +1,1289 @@
+        .execute()
+    # [END vector_length]
+    for res in result:
+        print(res)
```
We usually use the [`black` code formatter](https://pypi.org/project/black/) for our Python code. If you haven't used it, it might be good to run it over this file to see if it has any changes

Nothing really stands out to me as badly formatted, and I don't think using that formatting should be a requirement for samples. But it might be worth checking

daniel-sanche commented on snippets/firestore/firestore_pipelines.py:
```diff
@@ -0,0 +1,1289 @@
+            Field.of("rating").average().as_("avg_rating"),
+            groups=[Field.of("genre")]
+        ) \
+        .execute()
```
If you want to get rid of all the \\, IIRC think you could do something like this:

```
results = (
    client.pipeline()
    .collection("books")
    .aggregate(
        Field.of("rating").average().as_("avg_rating"),
        groups=[Field.of("genre")]
    ).execute()
)
```

(But test it first)

morganchen12 commented on snippets/firestore/firestore_pipelines.py:
```diff
@@ -0,0 +1,1289 @@
+
+# pylint: disable=invalid-name
+def pipeline_concepts():
+    # [START pipeline_concepts]
```
Yeah, I omitted `firestore_` where the filepath contained `firestore` somewhere. The snippets tool and the includecode tool both reference the filepath when importing statements so it'll still be findable as a Firestore snippet.

morganchen12 committed:
address feedback except formatting

morganchen12 committed:
run format

morganchen12 commented on snippets/firestore/firestore_pipelines.py:
```diff
@@ -0,0 +1,1289 @@
+        .execute()
+    # [END vector_length]
+    for res in result:
+        print(res)
```
Will defer to another PR

morganchen12 commented on snippets/firestore/firestore_pipelines.py:
```diff
@@ -0,0 +1,1289 @@
+    # [START pipeline_concepts]
+    pipeline = client.pipeline() \
+        .collection("cities") \
+        .where(Field.of("population").greater_than(100000)) \
```
Done here and throughout

morganchen12 committed:
remove confusing async bit

daniel-sanche reviewed (approved):
Consider adding more imports for Vector, and others that may be missing through the file

But overall LGTM

morganchen12 committed:
Add expression and input snippets, with errors

morganchen12 commented on snippets/firestore/firestore_pipelines.py:
```diff
@@ -0,0 +1,1753 @@
+    # [START cosine_distance]
+    from google.cloud.firestore_v1.pipeline_expressions import Field
+
+    sample_vector = Vector([0.0, 1.0, 2.0, 3.0, 4.0, 5.0])
```
Done.

daniel-sanche commented:
1. One of those is `from google.cloud import firestore`, and the other is `from firebase_admin import firestore`. I'm not super familiar with the `firebase_admin` wrapper, but I don't think those are fully equivalent

2. In the end, I think Array was actually implemented as a stand-alone class:

```
from google.cloud.firestore_v1.pipeline_expressions import Array
Array(["bar", Field.of("baz")])
```
It looks like there's an [out-dated docstring](https://github.com/googleapis/python-firestore/blob/4848fbe29fbed70f35067de694314dc91c2b68f1/google/cloud/firestore_v1/pipeline_expressions.py#L1853). I can fix that. 

Let me know if you think it's important to have `array` exposed on Expression as well though. I removed it in an effort to keep the API surface manageable, instead of exposing multiple ways of expressing the same thing. But I can add it back if that's more consistent with other languages.

4. array_get was a late addition, but it should be in the preview branch as of yesterday. Maybe try running again

morganchen12 committed:
add query explain snippet and fix lint errors

morganchen12 committed:
fix explainoptions snippet

morganchen12 commented on snippets/firestore/firestore_pipelines.py:
```diff
@@ -47,6 +20,19 @@
+        .where(filter=FieldFilter("capital", "==", True)) \
+        .execute(explain_options=ExplainOptions(analyze=False))
+    metrics = results.explain_metrics()
+    summary = metrics.plan_summary()
```
Should be fixed now. You can republish the docs on devsite and it'll pick up the fix.

morganchen12 committed:
add type and constant functions

morganchen12 committed:
Merge branch 'main' into mc/pipelines-snippets

morganchen12 committed:
add snippets except forceindex (not supported)

morganchen12 committed:
Add snippets for inserting data used by other snippets

morganchen12 committed:
add search and geo snippets

morganchen12 committed:
code review feedback

morganchen12 committed:
Merge branch 'main' into mc/pipelines-snippets

yvonnep165 committed:
feat(fcm): Enable `fid` and deprecate `token` for Send API (#951)

* add fid arg and token deprecate warning to Message class and add unit tests

* add fids to MulticastMessage and token deprecate warning and add unit tests

* Fix lint formatting error

* Position `tokens` first to guarantee backward compatibility for legacy positional arguments and add a unit test for it

* Extract multicast-to-message-list conversion logic into a private helper function

* Add docstring for the helper function

* Add unit tests for the async function, change deprecate message and address other review comments

* Fix error messages and MulticastMessage constructor and add mix types of fids and tokens

* Add integration tests for fid as argument

* Update integration tests error code for invalid fid target

lahirumaramba committed:
fix(deps): Added universe_domain override to MockGoogleComputeEngineCredential (#954)

Added a universe_domain property override to MockGoogleComputeEngineCredential in tests/testutils.py to prevent unexpected network requests to the GCE metadata server during test runs with newer versions of google-auth.

dependabot[bot] committed:
chore(deps): update pyjwt requirement from >=2.10.1 to >=2.12.1 (#948)

Updates the requirements on [pyjwt](https://github.com/jpadilla/pyjwt) to permit the latest version.
- [Release notes](https://github.com/jpadilla/pyjwt/releases)
- [Changelog](https://github.com/jpadilla/pyjwt/blob/master/CHANGELOG.rst)
- [Commits](https://github.com/jpadilla/pyjwt/compare/2.10.1...2.12.1)

---
updated-dependencies:
- dependency-name: pyjwt
  dependency-version: 2.12.1
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Yvonne Pan <103622026+yvonnep165@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): update google-cloud-firestore requirement (#947)

Updates the requirements on [google-cloud-firestore](https://github.com/googleapis/google-cloud-python) to permit the latest version.
- [Release notes](https://github.com/googleapis/google-cloud-python/releases)
- [Changelog](https://github.com/googleapis/google-cloud-python/blob/main/packages/google-cloud-documentai/CHANGELOG.md)
- [Commits](https://github.com/googleapis/google-cloud-python/compare/google-cloud-firestore-v2.23.0...google-cloud-firestore-v2.27.0)

---
updated-dependencies:
- dependency-name: google-cloud-firestore
  dependency-version: 2.27.0
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Yvonne Pan <103622026+yvonnep165@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): update pytest requirement from >=8.2.2 to >=8.4.2 (#946)

Updates the requirements on [pytest](https://github.com/pytest-dev/pytest) to permit the latest version.
- [Release notes](https://github.com/pytest-dev/pytest/releases)
- [Changelog](https://github.com/pytest-dev/pytest/blob/main/CHANGELOG.rst)
- [Commits](https://github.com/pytest-dev/pytest/compare/8.2.2...8.4.2)

---
updated-dependencies:
- dependency-name: pytest
  dependency-version: 8.4.2
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Yvonne Pan <103622026+yvonnep165@users.noreply.github.com>

morganchen12 committed:
Merge branch 'main' into mc/pipelines-snippets

morganchen12 committed:
Merge branch 'main' into mc/pipelines-snippets

morganchen12 committed:
add snippets for joins and explain (#973)

* add snippets for joins and explain

* fix snippet string
